### PR TITLE
Implementation of DropColumn for SQLite using migration transforms.

### PR DESCRIPTION
**Pull request by natemcmaster**
This is a first-cut of implementing a migration operation that requires table-rebuild.

Deliberate design choices:
 - Modifying the operation is provider specific and happens just before SQL generation
 - Migration API remains unchanged
 - Relies on the model to determine which columns are used in the table-rebuild.

Future improvements:
 - Add support for the many migration operations not available to SQLite
 - Allow users to opt-in for potentially-destructive table-rebuilds, instead of running them by-default.
 - Better logging/warning users of what is going to happen.

### CODE REVIEW

**Comment by bricelam** [comment]
Looks good, but I think we should flesh the feature out a bit more before merging into dev.

**Comment by natemcmaster** [comment]
Just added some more thorough testing. Also, these changes are local to SQLite only. 

As we discussed, we will build support for table-rebuilds first in SQLite, and afterwards find table-rebuild abstraction for other providers.

File: src/EntityFramework.Sqlite/Migrations/SqliteOperationTransformer.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        // TODO prevent multiple DropColumnOperations on the same table
+        private IList<MigrationOperation> TransformOperation(DropColumnOperation operation, IModel model)
+        {
+            Check.NotNull(model, nameof(model));
+
+            // TODO ensure this temporary table does not conflict with an existing table
+            var tempTableName = operation.Table + "_temp";
+
+            var createTableOperation = _differ.GetDifferences(null, model)
+                .FirstOrDefault(o => (o as CreateTableOperation)?.Name == operation.Table);
```
Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteOperationTransformer.cs]: We need a more efficient way to do this; creating all the operations and throwing away all but one is wasteful. Exposing `ModelDiffer.Add(IEntityType source, IEntityType target)` might be enough.

File: src/EntityFramework.Sqlite/Migrations/SqliteOperationTransformer.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+            if (!(createTableOperation is CreateTableOperation))
+            {
```
Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteOperationTransformer.cs]: It will always be a `CreateTableOperation`. Should this be `if (createTableOperation  == null)`

File: test/EntityFramework.Sqlite.FunctionalTests/MigrationTest.cs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    {
+        private readonly DbContextOptionsBuilder _options;
+        private readonly IServiceProvider _provider;
+        private readonly SqliteTestStore _testStore;
+
+        [Fact]
+        public void DropColumn_rebuilds_table()
+        {
+            _testStore.ExecuteNonQuery(@"CREATE TABLE Table1 (Id INT PRIMARY KEY, Col1, Col2); 
+INSERT INTO Table1 (Col1, Col2) Values('dropped value','preserved entry');");
```
Comment (bricelam) [inline test/EntityFramework.Sqlite.FunctionalTests/MigrationTest.cs]: Could use EF to setup the test. Create tables using `context.Database.EnsureCreated()` and insert as normal

File: test/EntityFramework.Sqlite.FunctionalTests/MigrationTest.cs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+            Migrate(up => { up.DropColumn("Col1", "Table1"); }, targetModel =>
+                {
+                    targetModel.Entity<UpdatedTableType>(b =>
+                        {
+                            b.Key(p => p.Id);
+                            b.Property(p => p.Id)
+                                .StoreGeneratedPattern(StoreGeneratedPattern.Identity);
```
Comment (bricelam) [inline test/EntityFramework.Sqlite.FunctionalTests/MigrationTest.cs]: Unnecessary. `StoreGeneratedPattern` doesn't influence much on SQLite yet.

File: test/EntityFramework.Sqlite.FunctionalTests/MigrationTest.cs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+            using (var context = CreateContext())
+            {
+                // Assert data moved
+                var all = context.Set<UpdatedTableType>().ToList();
+                Assert.Equal(1, all.Count);
+                Assert.Equal(1, all[0].Id);
+                Assert.Equal("preserved entry", all[0].Col2);
+
+                // Assert Primary Key generates new values
```
Comment (bricelam) [inline test/EntityFramework.Sqlite.FunctionalTests/MigrationTest.cs]: Asserting this doesn't seem related to the functionality you're testing. SQLite always generates new values for `INTEGER PRIMARY KEY` columns

File: test/EntityFramework.Sqlite.FunctionalTests/MigrationTest.cs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            {
+                var model = context.GetService<IMigrationModelFactory>().CreateModel(migration.BuildTargetModel);
+                var command = context.GetService<IMigrationSqlGenerator>().Generate(operations, model);
+
+                using (var transaction = context.Database.GetDbConnection().BeginTransaction())
+                {
+                    context.GetService<ISqlStatementExecutor>()
+                        .ExecuteNonQuery(context.Database.GetRelationalConnection(), transaction, command);
+                    transaction.Commit();
+                }
```
Comment (bricelam) [inline test/EntityFramework.Sqlite.FunctionalTests/MigrationTest.cs]: Any way to leverage parts of the Migrations pipeline for this?

File: test/EntityFramework.Sqlite.FunctionalTests/MigrationTest.cs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+                .BuildServiceProvider();
+
+            _options = new DbContextOptionsBuilder(new DbContextOptions<TestContext>());
+            _options.UseSqlite(_testStore.Connection);
+        }
+
+        private void AssertColumns(string tableName, List<string> columnNames)
+        {
+            var command = _testStore.Connection.CreateCommand();
+            command.CommandText = $"PRAGMA table_info({tableName});";
```
Comment (bricelam) [inline test/EntityFramework.Sqlite.FunctionalTests/MigrationTest.cs]: Would be neat to leverage the materializer and do something like `context.Columns.FromSql("PRAGMA table_info({0})", tableName).ToList();`

File: test/EntityFramework.Sqlite.Tests/Migrations/SqliteOperationTransformTest.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                    b.Property<string>("3").Column("col3");
+                    b.Key("3");
+                });
+            var transformer = new SqliteOperationTransformer(
+                new ModelDiffer(new SqliteTypeMapper(), new SqliteMetadataExtensionProvider()));
+            var actual = transformer.Transform(new[] { input }, model.Model);
+
+            Assert.Collection(actual, op1 =>
+                {
+                    var rename = (RenameTableOperation)op1;
```
Comment (bricelam) [inline test/EntityFramework.Sqlite.Tests/Migrations/SqliteOperationTransformTest.cs]: May get better failures with
```C#
var rename = Assert.IsType<RenameTableOperation>(op1);
```

**Comment by bricelam** [comment]
:shipit:

File: test/EntityFramework.Sqlite.FunctionalTests/MigrationTest.cs (hunk 3 above)
Comment (natemcmaster) [inline test/EntityFramework.Sqlite.FunctionalTests/MigrationTest.cs]: True, but this was much easier.

File: src/EntityFramework.Sqlite/Migrations/SqliteOperationTransformer.cs (hunk 1 above)
Comment (natemcmaster) [inline src/EntityFramework.Sqlite/Migrations/SqliteOperationTransformer.cs]: Will add as a TODO

File: test/EntityFramework.Sqlite.FunctionalTests/MigrationTest.cs (hunk 6 above)
Comment (natemcmaster) [inline test/EntityFramework.Sqlite.FunctionalTests/MigrationTest.cs]: No, but it would be nice to have an API for this in migrations so I don't have to recreate ApplyMigration()
